stdlib/public/core/NormalizedCodeUnitIterator.swift

```diff
@@ -349,63 +349,67 @@ struct _NormalizedUTF16CodeUnitIterator: IteratorProtocol {
 
   mutating func next() -> UInt16? {
     if segmentBufferCount == segmentBufferIndex {
+      if source.isEmpty {
+        return nil
+      }
       segmentBuffer = _FixedArray16<CodeUnit>(allZeros:())
       segmentBufferCount = 0
       segmentBufferIndex = 0
     }
-
-    if source.isEmpty
-    && segmentBufferCount == 0 {
-      // Our source of code units to normalize is empty and our buffers from
-      // previous normalizations are also empty.
-      return nil
-    }
+    
     if segmentBufferCount == 0 {
-      //time to fill a buffer if possible. Otherwise we are done, return nil
-      // Normalize segment, and then compare first code unit
-      if segmentHeapBuffer == nil,
-         let filled = source.tryFill(into: &normalizationBuffer)
-      {
-        if let count = _tryNormalize(
-          _castOutputBuffer(&normalizationBuffer,
-          endingAt: filled),
-          into: &segmentBuffer
-        ) {
-          segmentBufferCount = count
-        } else {
-          segmentBufferCount = normalizeWithHeapBuffers()
-        }
-      } else {
-        segmentBufferCount = normalizeWithHeapBuffers()
-      }
+      segmentBufferCount = normalizeFromSource()
     }
 
     if segmentBufferIndex < segmentBufferCount {
+      defer { segmentBufferIndex += 1 }
       if _slowPath(segmentHeapBuffer != nil) {
-        _sanityCheck(segmentBufferIndex < segmentHeapBuffer!.count)
-        let index = segmentBufferIndex
-        segmentBufferIndex += 1
-        return segmentHeapBuffer![index]
+        return segmentHeapBuffer![segmentBufferIndex]
       } else {
-        let index = segmentBufferIndex
-        segmentBufferIndex += 1
-        return segmentBuffer[index]
+        return segmentBuffer[segmentBufferIndex]
       }
     } else {
       return nil
     }
   }
   
-  mutating func normalizeWithHeapBuffers() -> Int {
+  mutating func normalizeFromSource() -> Int {
+    if segmentHeapBuffer == nil,
+       let filled = source.tryFill(into: &normalizationBuffer)
+    {
+      if let count = _tryNormalize(
+        _castOutputBuffer(&normalizationBuffer,
+        endingAt: filled),
+        into: &segmentBuffer
+      ) {
+        return count
+      } else {
+        return normalizeWithHeapBuffers(filled)
+      }
+    } else {
+      return normalizeWithHeapBuffers()
+    }
+  }
+  
+  mutating func normalizeWithHeapBuffers(
```

Comment:
This has gotten pretty complicated. Can you document it and add comments explaining the behavior?